### ISSUE TITLE
64-bit libatomic not build with support for 16-byte compare/exchange

### ISSUE THREAD

**Issue by ulatekh** [msys2/MINGW-packages#13831]
### Description

libatomic's __atomic_compare_exchange_n() can provide support for the cmpxchg16b instruction, but that has to be set at build time.
From single-stepping through the generated assembly, it appears it's not being compiled in, which means it defaults to using a mutex, and that is terrible for performance. (I stepped from libatomic into winpthread.)
All reasonably-available x86_64 processors these days support cmpxchg16b.
Can you please enable support for it in libatomic, at the very least in the 64-bit build?

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

Microsoft Windows [Version 10.0.19044.2130]

### MINGW environments affected

- [X] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

### Expected behavior

Using the cmpxchg16b instruction

### Actual behavior

Uses a mutex

### Repro steps

Should be easy to verify

**Comment by lazka** [comment]
Fine with me. Any idea for how to enable this?

**Comment by jeremyd2019** [comment]
Maybe it just needs to be rebuilt post `-march` change?

**Comment by lazka** [comment]
That has already happened, afaics

**Comment by jeremyd2019** [comment]
I started out down that path too, but it seems libatomic is part of gcc while libatomic_ops is different.  But that is concerning, indicating that gcc is not entirely happy with cmpxchg16b somehow.

**Comment by Biswa96** [comment]
Reverting `-march=nocona -msahf` to `-march=x86-64` fixes the build.

**Comment by ulatekh** [comment]
The compiler flag -mcx16 has been bandied about, but if -march=x86-64 does it, that's fine.
The other machine on which I've single-stepped through the assembly was running Ubuntu 18, and it used cmpxchg16b.

**Comment by lazka** [comment]
> I get the following error while re-building libatomic
> 
> ```
> ld.exe: .libs/atomic_ops_stack.o:atomic_ops_stack.c:(.text+0x8d): undefined reference to `__atomic_compare_exchange_16'
> ```

This is just missing a `LDFLAGS+=" -latomic"` I think, so not related, and atomic_ops != libatomic.

**Comment by jeremyd2019** [comment]
`-march=x86-64` turns *off* support for `comxchg16b`.  We recently enabled `march` flags that should allow it to be used, but there seem to be some issues to figure out.

**Comment by lhmouse** [comment]
Looks like something changed since GCC 7:
```c
typedef unsigned uint128 __attribute__((mode(TI), aligned(16)));

int
test_cmpxchg16b(uint128* mem, uint128* cmp, uint128 val)
  {
    return __atomic_compare_exchange_16(mem, cmp, val,
           0, __ATOMIC_ACQ_REL, __ATOMIC_ACQUIRE);
  }
```
GCC 7 outputs:
```asm
test_cmpxchg16b:
        sub     rsp, 8
        mov     r9d, 2
        mov     r8d, 4
        call    __atomic_compare_exchange_16
        add     rsp, 8
        movzx   eax, al
        ret
```
GCC 6 outputs:
```asm
test_cmpxchg16b:
        push    rbx
        mov     rax, QWORD PTR [rsi]
        mov     rbx, rdx
        mov     rdx, QWORD PTR [rsi+8]
        lock cmpxchg16b XMMWORD PTR [rdi]
        mov     r10, rdx
        sete    dl
        je      .L2
        mov     QWORD PTR [rsi], rax
        mov     QWORD PTR [rsi+8], r10
.L2:
        movzx   eax, dl
        pop     rbx
        ret
```

Link: https://gcc.godbolt.org/z/cdesffae1

**Comment by jeremyd2019** [comment]
So GCC >= 7 needs libatomic's implementation to use `cmpxchg16b`, and you have to live with the function call overhead now?  Or does that mean GCC >= 7 doesn't want to generate `lock cmpxchg16b` anymore?

**Comment by lhmouse** [comment]
I was given <this https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80878>  on #gcc on OFTC.

Apparently, from GCC 7 to 12, no GCC version generates `CMPXCHG16B`, even when that's explicitly requested; only Clang does.

**Comment by lhmouse** [comment]
Another option is to use the deprecated `__sync_val_compare_and_swap()` which will generate `CMPXCHG16B`.

I was told that

> [11:44:22] <pinskia> no that is correct behavior
[11:45:08] <pinskia> clang is broken ....
[11:45:36] <pinskia> but clang is broken for other reasons on x86_64 because it does not implement the ABI correctly
[11:45:43] <pinskia> but they say they don't care ...

suggesting that `-latomic` is actually required, even when `-mcx16` is given.

**Comment by lhmouse** [comment]
Another option is to use the deprecated `__sync_val_compare_and_swap()` which will generate `CMPXCHG16B`.

I was told that

> [11:43:21] <lh_ideapad> does it mean that, if we have `-mcx16` when building GCC and its libs, there would be no `__atomic_*_16` in libatomic?
[11:43:26] <pinskia> oh arch
[11:43:32] <pinskia> o
[11:43:34] <pinskia> no
[11:43:40] <pinskia> it will always call it
[11:43:44] <lh_ideapad> no no  apologies for the confusion.
[11:43:45] <lh_ideapad> https://github.com/msys2/MINGW-packages/issues/13831
[11:44:07] <lh_ideapad> MSYS2 also uses pacman as its package manager.
[11:44:22] <pinskia> no that is correct behavior
[11:45:08] <pinskia> clang is broken ....
[11:45:36] <pinskia> but clang is broken for other reasons on x86_64 because it does not implement the ABI correctly
[11:45:43] <pinskia> but they say they don't care ...

suggesting that `-latomic` is actually required, even when `-mcx16` is given.

**Comment by jeremyd2019** [comment]
OK, sounds like this is intended behavior for upstream, rather than something broken in the build/packaging, and there is an upstream bug about it.

**Comment by lhmouse** [comment]
> OK, sounds like this is intended behavior for upstream, rather than something broken in the build/packaging, and there is an upstream bug about it.

I think that we should patch GCC locally. Such a decision is casual, arbitrary and irresponsible.

**Comment by jeremyd2019** [comment]
> I think that we should patch GCC locally. Such a decision is casual, arbitrary and irresponsible.

I think that should be a last resort, after careful consideration of potential impacts (there may be code elsewhere in GCC's ecosystem that relies on this decision and passes read-only memory in for an atomic read operation).  Hopefully something comes of the discussions upstream.

**Comment by FrankHB** [comment]
> it does not implement the ABI correctly

Which rule as *the* ABI?

**Comment by ulatekh** [comment]
FYI, I'm trying to get libatomic built so it uses `cmpxchg16b` .
I'm not thrilled with the extra function call, but that's how it works under Ubuntu 18, also.
The issue is that libatomic under MinGW isn't using `cmpxchg16b`, but a mutex.
I wasn't expecting `__atomic_compare_exchange_16()` to emit `cmpxchg16b` directly, but to use `cmpxchg16b` internally.

**Comment by lhmouse** [comment]
> FYI, I'm trying to get libatomic built so it uses `cmpxchg16b` .
> 
> I'm not thrilled with the extra function call, but that's how it works under Ubuntu 18, also.
> 

On Unubtu, GCC still generates a call to `__atomic_compare_exchange_16` even with very new arch options, such as `-march=rocketlake`.

I suspect Ubuntu people have some patches for GCC on this. It would also be helpful to have a look at their configuration options.

**Comment by ulatekh** [comment]
@lhmouse : Yes, but inside `__atomic_compare_exchange_16`, Ubuntu uses `cmpxchg16b`, whereas in current MinGW, it doesn't, and ends up going inside libpthread to lock a mutex.
